src/compiler/checker.ts

```diff
@@ -22966,19 +23218,44 @@ namespace ts {
                   }
 
-                  In the first clause (case `number` and `string`) the
-                  implied type is number | string.
+                  In the first clause (case `number` and `string`) the implied type is number | string.
 
                   In the default clause we de not compute an implied type.
 
-                  In the third clause (case `number` and `boolean`)
-                  the naive implied type is number | boolean, however
-                  we use the type facts to narrow the implied type to
-                  boolean. We know that number cannot be selected
+                  In the third clause (case `number` and `boolean`) the naive implied type is number | boolean, however
+                  we use the type facts to narrow the implied type to boolean. We know that number cannot be selected
                   because it is caught in the first clause.
                 */
+                const expr = (<TypeOfExpression>switchStatement.expression).expression;
                 const impliedType = getTypeWithFacts(getUnionType(clauseWitnesses.map(text => getImpliedTypeFromTypeofGuard(type, text) || type)), switchFacts);
-                return getTypeWithFacts(mapType(type, narrowUnionMemberByTypeof(impliedType)), switchFacts);
+                if (isMatchingReference(reference, expr)) {
+                    if (hasDefaultClause) {
+                        return filterType(type, t => (getTypeFacts(t) & switchFacts) === switchFacts);
+                    }
+                    return getTypeWithFacts(mapType(type, narrowUnionMemberByTypeof(impliedType)), switchFacts);
+                }
+                const impliedUndefined = !(switchFacts & 0b111111 && !(switchFacts & TypeFacts.EQUndefined) || (switchFacts & TypeFacts.NEUndefined));
+                const propertyTypeArray = getPropertyTypesFromTypeAccordingToExpression(<UnionType>type, expr);
+
+                if (!propertyTypeArray) {
+                    return type;
+                }
+                const markSet = new Set<TypeId>();
+
+                propertyTypeArray.forEach(propertyType => {
+                    if (!propertyType.finalType && !propertyType.containsEffectiveOptionalChain) { return; }
+                    const reachableFinalType = propertyType.finalType;
+                    if (propertyType.containsEffectiveOptionalChain && impliedUndefined) {
+                        markSet.add(propertyType.typeId);
+                    }
+                    else if (!!reachableFinalType) {
+                        const propertyTypeFacts = getTypeFacts(reachableFinalType);
+                        if (!hasDefaultClause ? propertyTypeFacts & switchFacts : ((propertyTypeFacts & switchFacts) === switchFacts)) {
+                            markSet.add(propertyType.typeId);
+                        }
+                    }
+                });
+                return filterType(type, (t) => markSet.has(t.id));

```

Comment:
I remember now; I think the reason I changed `NarrowDeepPropertyInfo` to be a map was I thought it would let me get rid of all these `markSet` Sets and the separate `propertyTypeArray.forEach`, and do simple, cheap lookups directly in `filterType`. But I discovered the other bugs before getting there, so I’m not 100% sure how it would have worked out. But it would be nice to simplify this. There are a lot of separate pieces and separate iterations that make it verbose and hard to follow.